src/core/layertree/qgslayertreemodel.cpp

```diff
@@ -1299,10 +1303,41 @@ QList<QgsLayerTreeModelLegendNode*> QgsLayerTreeModel::layerLegendNodes( QgsLaye
 
 void QgsLayerTreeModel::legendInvalidateMapBasedData()
 {
+  // we have varying icon sizes, and we want icon to be centered and
+  // text to be left aligned, so we have to compute the max width of icons
+  //
+  // we do that for nodes who share a common parent
+  //
+  // we do that here because for symbols with size defined in map units
+  // the symbol sizes changes depends on the zoom level
+
+  QList<QgsSymbolV2LegendNode*> symbolNodes;
+  QMap<QString, int> widthMax;
```

Comment:
widthMax should work per-layer - otherwise it may overwrite keys among different layers (especially root key which is empty string)